vcd/datasource_vcd_nsxt_app_port_profile.go

```diff
@@ -85,19 +101,77 @@ func datasourceVcdNsxtAppPortProfileRead(ctx context.Context, d *schema.Resource
+		inheritedVdcField := vcdClient.Vdc
+		contextId, err := pickAppPortProfileContextFilterByPriority(vcdClient, d, inheritedVdcField, dataSourceVdcField, contextIdField)
+		if err != nil {
+			return diag.Errorf("error identifying correct context filter: %s", err)
+		}
+		queryParams.Add("filter", fmt.Sprintf("name==%s;scope==%s;_context==%s", name, scope, contextId))
+	// For PROVIDER scoped App Port Profiles context_id of Network Provider can be specified
+	case strings.EqualFold(scope, types.ApplicationPortProfileScopeProvider) && contextIdField != "":
+		queryParams.Add("filter", fmt.Sprintf("name==%s;scope==%s;_context==%s", name, scope, contextIdField))
+	// Deprecated field 'nsxt_manager_id' can be specified as context for PROVIDER scoped App Port Profiles
+	case strings.EqualFold(scope, types.ApplicationPortProfileScopeProvider) && nsxtManagerId != "":
+		queryParams.Add("filter", fmt.Sprintf("name==%s;scope==%s;_context==%s", name, scope, nsxtManagerId))
+	default:
+		// For "SYSTEM" or "PROVIDER" scoped Application Port Profiles context can be ignored.
+		// * For "SYSTEM" this is correct behavior
+		// * For "PROVIDER" it can match App Port Profiles when multiple NSX-T Managers are
+		// configured, but this is left for backwards compatibility
+		//
+		// TODO V4 - remove support for PROVIDER scope without `context_id` field
+		queryParams.Add("filter", fmt.Sprintf("name==%s;scope==%s", name, scope))
+	}
+
+	allAppPortProfiles, err := org.GetAllNsxtAppPortProfiles(queryParams, scope)
 
-	appPortProfile, err := org.GetNsxtAppPortProfileByName(name, scope)
 	if err != nil {
-		return diag.Errorf("error getting NSX-T Application Port Profile with Name '%s' (scope '%s'): %s",
-			name, scope, err)
+		return diag.Errorf("error retrieving Application Port Profiles: %s", err)
+	}
+
+	if len(allAppPortProfiles) == 0 {
+		return diag.Errorf("%s NSX-T Application Port Profile not found", govcd.ErrorEntityNotFound)
 	}
 
+	if len(allAppPortProfiles) > 1 {
+		return diag.Errorf("Expected exactly one NSX-T Application Port Profile. Got '%d'", len(allAppPortProfiles))
+	}
+	appPortProfile := allAppPortProfiles[0]
+
 	err = setNsxtAppPortProfileData(d, appPortProfile.NsxtAppPortProfile)
 	if err != nil {
-		return diag.Errorf("error reading NSX-T Application Port Profile: %s", err)
+		return diag.Errorf("error storing NSX-T Application Port Profile schema: %s", err)
 	}
 
 	d.SetId(appPortProfile.NsxtAppPortProfile.ID)
 
 	return nil
 }
+
+// pickAppPortProfileContextFilterByPriority will evaluate 3 fields - 'context_id', 'vdc' in
+// resource and 'vdc' in provider section. It will pick the right one based on priority:
+// * Priority 1 -> 'context_id' field
+// * Priority 2 -> 'vdc' field in data source
+// * Priority 3 -> 'vdc' field inherited from provider configuration
+func pickAppPortProfileContextFilterByPriority(vcdClient *VCDClient, d *schema.ResourceData, inheritedVdcField, dataSourceVdcField, contextIdField string) (string, error) {
```

Comment:
`dataSourceVdcField ` is not used

Reply:
Indeed, as well as `inheritedVdcField`. They are not used in the latest iteration as I rely on the well tested code `vcdClient.GetOrgAndVdcFromResource`. Removed both parameters.